### ISSUE TITLE
Create option to reuse an existing ALB instead of creating a new ALB per Ingress

### ISSUE THREAD

**Issue by julianvmodesto** [kubernetes-sigs/aws-load-balancer-controller#298]
I read in this comment https://github.com/coreos/alb-ingress-controller/issues/85#issuecomment-299504058 that host-based routing was released for AWS ALBs shortly after ALB Ingress Controller was released.

It would be pretty cool to have an option to reuse an ALB for an Ingress via annotation -- i'd be interested in contributing towards this, but I'm not sure what's needed to make this feasible.

**Comment by pperzyna** [comment]
@bigkraig  Any update?

**Comment by mwelch-ptc** [comment]
Wait... I guess I missed this in reading the documentation. Are you saying that every Ingress created deploys it's own ALB? So for our 60 or so ingresses we'd end up with 60 ALB's? What about different host names within the same ingress? Does that at least reuse the same ALB?

**Comment by patrickf55places** [comment]
@mwelch-ptc That is correct. There is a 1-to-1 mapping of `Ingress` resources to ALBs, even if host names are the same.

**Comment by kurtdavis** [comment]
Seems to be fairly costly.  We have looked at other solutions due to this issue.

**Comment by bigkraig** [comment]
What are your guys thoughts on how to prioritize the rules if a single ALB spans ingress resources and potentially event namespaces? I can see where in larger clusters multiple teams may accidentally take the same path.

**Comment by vainu-arto** [comment]
> What is everyones thoughts on how to prioritize the rules if a single ALB spans ingress resources and potentially event namespaces? I can see where in larger clusters multiple teams may accidentally take the same path.

This is a general Kubernetes ingress issue, not specific to this ingress controller. I think the discussion of this should be had in a more general forum instead of an issue against this controller.

**Comment by whithajess** [comment]
Im tempted to say that this is not a general Kubernetes ingress issue.

Existing load balancers supported by Kubernetes are Layer 4 - And are supported by Ingress Controllers that do the Layer 7 (This means they can use 1 load balancer and then deal with layer 7 when it gets into the cluster) 

ALB is Layer 7 and is dealing with it before it gets to Kubernetes, so we cannot assume they are going to change for this use case.

**Comment by spacez320** [comment]
We can already generally do this by having a singular ingress resource, although it makes whatever deployment scheme you're using for Kubernetes have to adjust to that. It's also worth pointing out that in Kubernetes Ingress documentation, it literally states:

> An Ingress allows you to keep the number of loadbalancers down to a minimum.

I think it would be really nice to have the ability to do this in a nice way.

**Comment by bigkraig** [comment]
@spacez320 I read that as that you can have an ingress with multiple services behind it, so a single load balancer for many services as opposed to a load balancer per service.

There is still the issue that the IngressBackend type does not have a way to reference a service in another namespace. I think until the ingress resource spec is changed, there isn't a correct way of implementing something like this.

**Comment by patrickf55places** [comment]
@bigkraig I don't think this issue should be closed. The issue isn't about having a single `Ingress` resource that can span multiple namespaces. It is about having multiple `Ingress` resources (possible across different namespaces, but not necessarily) that all use the same AWS application load balancer.

**Comment by bigkraig** [comment]
@patrickf55places got it, within the namespace is possible with the spec but I am still unsure how we would organize the routes or resolve conflicts.

**Comment by spacez320** [comment]
@bigkraig Well, I think it's both, and I think that's what @patrickf55places meant by saying "possible across different namespaces, but not necessarily". We should be able to define an Ingress anywhere and share an Amazon load balancer, I think.

I understand if there's limitations in the spec, though. Should someone go out and try to raise this issue with the wider community? Is that possibly already happening?

**Comment by natefox** [comment]
What about using something similar to how nginx ingress handles it?
https://github.com/nginxinc/kubernetes-ingress/tree/master/examples/mergeable-ingress-types

> Multiple minions can be applied per master as long as they do not have conflicting paths. If a conflicting path is present then the path defined on the oldest minion will be used.

**Comment by joegoggins** [comment]
I was glad to find this GitHub issue and also bummed that it seems like it will be a long time before this will get implemented. It smells like there is a lot of complexity associated with the change and potentially not resources to dig into it. I'm assuming it will be many months and thus our engineering team is going to switch our technical approach to use a different load balancing ingress strategy with AWS costs that scale economically in-line with our needs.  If that assessment feels wrong, please let me know.

**Comment by jakubkulhan** [comment]
I've created another ingress controller that combines multiple ingress resources into a new one => https://github.com/jakubkulhan/ingress-merge 

Partial ingresses are annotated with `kubernetes.io/ingress.class: merge`, merge ingress controller processes them and outputs a new ingress annotated with `kubernetes.io/ingress.class: alb`, then ALB ingress controller takes over and creates single AWS load balancer.

**Comment by kainlite** [comment]
I'm currently using ingress-merge and while it works I'm having issues with the health checks as the services that I'm exposing do different things by default and we don't have a standard health check url for all microservices, do you have a solution for this issue?, I think the limitation comes from aws-alb-ingress-controller rather than ingress-merge, but if there is a way to have different health checks that would be awesome. Thanks everyone for your effort.

**Comment by kainlite** [comment]
@fygge on slack gave me the answer: 

> You can put the health check annotation on the service instead of on the ingress resource. Thereby having one health check per target group / service.

Tested and works ok.

**Comment by mlsmaycon** [comment]
> What about using something similar to how nginx ingress handles it?
> https://github.com/nginxinc/kubernetes-ingress/tree/master/examples/mergeable-ingress-types
> 
> > Multiple minions can be applied per master as long as they do not have conflicting paths. If a conflicting path is present then the path defined on the oldest minion will be used.

ALB does that with [listener rules priority](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/listener-update-rules.html), where  a new rule get a higher priority than existing rules (excluded default rule). Problem will be if you set a priority number that conflicts with an existing rule. 

Maybe a new kind of ingress controller is in call here, something that with the Ingress object, controls only target groups and listener rules and attach then to a ALB created at the controller configuration(or at first object request). This is what this issue is asking with the current controller, but for larger organization, this could bring complexity with path or host-header rules, causing problems with overlapping Ingress objects.

**Comment by tdmalone** [comment]
Relevant: https://github.com/kubernetes-sigs/aws-alb-ingress-controller/issues/914

**Comment by jzangari** [comment]
I'm working on a multi tenant application where each customer will have their own namespace running a copy of the application. We'd still like to use a single ingress point for the cluster using an ALB to route across namespaces rather than and ALB in each namespace.  So I'm bumping this before it goes competely stale.

**Comment by nadundesilva** [comment]
I am working on a project where we have many helm charts and all of them contain their own ingress resources. It doesn't make sense to bring everything together into a single ingress resource because of the helm chart level grouping of resources.

It would be really nice if we can share a single ALB between ingresses.

**Comment by kuznero** [comment]
Sad if it is going to go unnoticed

**Comment by oceaneLonneux** [comment]
I am currently implementing an ALB with an EKS cluster, we have dozens of services and multiple environments, therefore different ALB is financially not possible. 

I'm a bit surprised this hasn't been more mentioned, as to me it is a true problem. I will investigate and hopefully, will come back with how we do. Will try the merge mentioned earlier!

**Comment by hectoralicea** [comment]
I have a similar situation where my customer wants to create an ALB outside of EKS to load the certificate with short fixed host and alternate long random ALB host name to the ALB.  Then when the service is created in EKS using helm it would be exposed using an internal ELB.  The idea was to then link the ALB to the ELB.  This is proving out to be much more difficult than I envisioned.  Any ideas would be appreciated.

**Comment by M00nF1sh** [comment]
@email2liyang You can try out [v1.2.0-alpha.1](https://github.com/kubernetes-sigs/aws-alb-ingress-controller/releases/tag/v1.2.0-alpha.1), which supports multiple Ingress reuse same ALB. (**Note: it's not production ready and not compatible with older version yet**)

**Comment by email2liyang** [comment]
@M00nF1sh , Thanks for the quick reply, I'm using ingress controller in production env, I'm very happy when I discover this, because I could combine the ALB ingress controller and ACM out of box to provide HTTPS support, and migrate from[ingress-nginx](https://kubernetes.github.io/ingress-nginx/) to alb ingress controller,  I even write a [blog](https://www.vipmind.me/infra/kube/expose-webapplicant-with-aws-alb-ingress-controller-and-aws-certifacate-manager-over-https.html)  to record my process on how to do it, but in the end, the issue that "ALB ingress could not reference services across namespace " becomes the last hurdle of it and I do not want to create one ALB for one service or at least one ALB for one namespaces

After v1.2.0 is production ready, I will definitely try ALB ingress controller again

Thanks

**Comment by rparsonsbb** [comment]
> I have a similar situation where my customer wants to create an ALB outside of EKS to load the certificate with short fixed host and alternate long random ALB host name to the ALB. Then when the service is created in EKS using helm it would be exposed using an internal ELB. The idea was to then link the ALB to the ELB. This is proving out to be much more difficult than I envisioned. Any ideas would be appreciated.

@hectoralicea it is not possible to point an alb to an elb this is a limitation of AWS functionality.

https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-target-groups.html#target-type

**Comment by thunder-spb** [comment]
Any updates here?

**Comment by styk-tv** [comment]
I will go a bit off topic here > ALB to ELB (Classic). I have been using ELB-C for the past few years and running across very many applications. Single ingress, wildcard certificates terminated at ELB and multi-node ingress. I've have tried HAProxy, Traefik and Nginx (default) with success. If you really push for (differently priced ALB) then its really your issue, if you just want to solve the problem and move on, going back to ELB-Classic may not be a bad idea. It works! Unless you have some extreme examples of why you need multiple cloud load balancers.  There are some issues always with a small amount of applications (reverse proxy, session stickyness, oauth redirect) but for the most part you can still get them to work on a single LB. At the end its a cost analysis issue. If you want to reduce 5 to 1 and your time is expensive then don't do it. If you have 1000 load balancers and you can reduce it to 3, it might be worth a while.  And even if you create ALB outside of Ingress type automation, then you can still map all your services through ALB, just don't specify the loadbalancer type and manually point the ALB at all ingress nodeports.

**Comment by ratulbasak** [comment]
Any update on this issue?

**Comment by MayconRitzmann** [comment]
I have a similar problem.

Each ingress created in the eks cluster, an alb in my environment goes up.

Anyone else with this problem?

**Comment by brunojcm** [comment]
@ritzmann94 this is not a "problem", it's a "lack of feature", I'd say, and that's all this issue is all about. Everyone here is on the same boat.

I personally have been using a NLB with ingress-nginx, all my ingresses objects get merged on a single nginx and share the same NLB. I know it's now officially supported, but I got sick of waiting for ALB support.

**Comment by ritzmann94** [comment]
Hello @brunojcm, when you said: "I know that it is now officially supported, but I was tired of waiting for ALB's support".

I didn't understand it very well, at least in the official documentation we don't have any reference to use just 1 ALB.

**Comment by brunojcm** [comment]
> Hello @brunojcm, when you said: "I know that it is now officially supported, but I was tired of waiting for ALB's support".
> 
> I didn't understand it very well, at least in the official documentation we don't have any reference to use just 1 ALB.

Sorry, I got auto-corrected, I meant "not", not "now. I was talking about NLB support, the one I ended up using, no ALB. ALB is supported, just not sharing the same ALB for multiple ingresses. NLB is still in alpha/beta (sorry if I'm not up-to-date here).

**Comment by ritzmann94** [comment]
Understand, I need to deploy AWS WAF in my environment and being in NLB with ingress-nginx is not supported.

Let's wait for this new feature to come.

**Comment by msolimans** [comment]
Any updates?

**Comment by bittelc** [comment]
This is an issue that should be made very very clear on the ALB ingress controller documentation, it currently is not. If I had known this was the case, I would have pivoted to a different ingress a lot earlier in my PoC.
Now I'm frustrated.

**Comment by vprus** [comment]
For avoidance of doubt, this issue is actually fixed in a beta release specifically docker.io/amazon/aws-alb-ingress-controller:v1.2.0-alpha.1 and I have a dozen of ingresses sharing single ALB. But, that beta release made made a year ago. It would be very nice to get some clarity whether it's coming in any official form.

**Comment by kirkdave** [comment]
Would be great to get this feature either merged into the latest code or revitalise v1.2

I would love to use this feature, but also want to be able to use IAM Role for Service Accounts in EKS and having tested the tag v1.0.0-alpha.1 that support hasn't been merged in (works great on v1.1.8).

**Comment by dmanchikalapudi** [comment]
> For avoidance of doubt, this issue is actually fixed in a 1.2 alpha release, specifically docker.io/amazon/aws-alb-ingress-controller:v1.2.0-alpha.1 and I have a dozen of ingresses sharing single ALB. But, that alpha release was made a year ago. It would be very nice to get some clarity whether it's coming in any official form.

How did you get that to work? Can you share the ingress definitions for a couple applications? Are you getting it to work by specifying same ingress name but define a different rule in both? Also, are you using a package / deployment manager like Helm 3? I believe it has validations to not deploy an existing resource (The request to create the ingress does not even get to k8s)

**Comment by vprus** [comment]
Here's a complete helm template used to define ingress for one particular application. 
```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: {{ .Release.Name }}-ingress
  annotations:
    kubernetes.io/ingress.class: alb
    alb.ingress.kubernetes.io/scheme: internal
    alb.ingress.kubernetes.io/group.name: analytics
  labels:
    app: {{ .Release.Name }}
spec:
  rules:
    - host: {{ .Release.Name }}.somecompany.a
      http:
        paths:
          - path: /*
            backend:
              serviceName: {{ .Release.Name }}-jobmanager-external
              servicePort: 8081
```
We use Helm 3, but exact same definition worked in Helm 2 too. The key part is 'group.name' metadata above. Two helm releases that use this template result in two ingress objects, with different names, and then in two target groups used by a single ALB. The deployment of alb-ingress-controller is basically using default options, except for
```
image: docker.io/amazon/aws-alb-ingress-controller:v1.2.0-alpha.1
```

**Comment by dmanchikalapudi** [comment]
> Here's a complete helm template used to define ingress for one particular application.
> 
> ```
> apiVersion: extensions/v1beta1
> kind: Ingress
> metadata:
>   name: {{ .Release.Name }}-ingress
>   annotations:
>     kubernetes.io/ingress.class: alb
>     alb.ingress.kubernetes.io/scheme: internal
>     alb.ingress.kubernetes.io/group.name: analytics
>   labels:
>     app: {{ .Release.Name }}
> spec:
>   rules:
>     - host: {{ .Release.Name }}.somecompany.a
>       http:
>         paths:
>           - path: /*
>             backend:
>               serviceName: {{ .Release.Name }}-jobmanager-external
>               servicePort: 8081
> ```
> 
> We use Helm 3, but exact same definition worked in Helm 2 too. The key part is 'group.name' metadata above. Two helm releases that use this template result in two ingress objects, with different names, and then in two target groups used by a single ALB. Other applications define their ingresses in a similar way, and also share the same ALB. The deployment of alb-ingress-controller is basically using default options, except for
> 
> ```
> image: docker.io/amazon/aws-alb-ingress-controller:v1.2.0-alpha.1
> ```

Thanks ! I will give it a try and confirm if its working for our use-cases. 

Btw - https://github.com/kubernetes-sigs/aws-alb-ingress-controller/issues/914 states that this is **not** production ready and should not be used. Any idea when it will be available as an official release?

**Comment by ffjia** [comment]
> IAM Role for Service Accounts
@kirkdave did you make the ALB ingress IRSA work in EKS?

**Comment by kirkdave** [comment]
@ffjia It works with IRSA if you either build the Docker image from the branch or, as I did, use the image that @M00nF1sh created in #914 - m00nf1sh/aws-alb-ingress-controller:v1.2.0-alpha.2

**Comment by rajeshwrn** [comment]
I was came across the same kind of requirement.

Have to deploy 20+ services in kubernetes in AWS Fargate profile. Since the fargate is not support NLB as of now the only option is ALB. But each deployment created new Alb. This needs 20+ public ips and also more cost on alb.

I achieved the solution with two ingress controller, alb ingress and nginx ingress.

Nginx will be the target for alb with port 80. And application services running in cluster with different port and namespaces will communicate nginx.

I have documented my solution. I think it will help your requirement.

https://github.com/rajeshwrn/alb-nginx-controller

**Comment by MXClyde** [comment]
Any timelines on when this functionality will be part of a stable release?

**Comment by astrived** [comment]
@MXClyde we are doing the final phase of testing for the new aws-loadbalancer-controller and the RC image is here https://github.com/kubernetes-sigs/aws-alb-ingress-controller/releases/tag/v2.0.0-rc5. You can find details here https://github.com/kubernetes-sigs/aws-alb-ingress-controller/tree/v2_ga. Stay tuned for What's new coming soon !

**Comment by astrived** [comment]
We have published the v2.0.0 release at https://github.com/kubernetes-sigs/aws-load-balancer-controller/releases/tag/v2.0.0 The documents have been updated https://docs.aws.amazon.com/eks/latest/userguide/alb-ingress.html and https://docs.aws.amazon.com/eks/latest/userguide/load-balancing.html and the blog post is here https://aws.amazon.com/blogs/containers/introducing-aws-load-balancer-controller/

**Comment by haalcala** [comment]
Can you guys improve your documentation system like how mongodb or elasticsearch does it? Like in a way that I can select which version of the library/system documentation I want to view?  I mean, can you not overwrite the published url with the latest one, coz I have to maintain some old versions, coz it gets confusing trying remember what I did before (with 1.1.4 for example) and looking at version 2.

**Comment by angapov** [comment]
It seems this is still not implemented. Any plans for supporting it in future?

**Comment by GoodMirek** [comment]
I am quite sure this feature is already implemented, I used it recently.
Look at https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.4/guide/ingress/annotations/#ingressgroup

**Comment by angapov** [comment]
Oops, sorry, my bad. I actually thought about reusing existing ALB created outside of Ingress controller (sharing the ALB between multiple EKS clusters). But nevermind.

**Comment by Tzrlk** [comment]
I've been thinking to try manually adding the tags that the ingress controller uses to identify the ALBs it creates across restarts, so the ALB can be created outside kubernetes and simply controlled by the controller.
The benefits of this approach are that I wouldn't have to give the controller permission to create ALBs, only target groups, etc. and could reduce the complexity of my k8s config significantly by configuring the obvious stuff up-front (http -> https redir, ssl cert, etc.)
It'd be great if the devs can confirm whether that approach is currently possible, otherwise when I get back to that area of my current project, I'll give it a spin, and report back.

**Comment by vprus** [comment]
@Tzrlk : just to clarify, you can configure these things with ALB controller. The `alb.ingress.kubernetes.io/certificate-arn` annotation allows to set SSL certificate, the `alb.ingress.kubernetes.io/ssl-policy` one sets SSL policy and to redirect HTTP you do something like
```
alb.ingress.kubernetes.io/listen-ports: '[{"HTTP": 80}, {"HTTPS": 443}]'
alb.ingress.kubernetes.io/ssl-redirect: '443'
```

I agree that reusing an existing balancer might be a nice feature, but most of configurability was added to the ALB controller.

**Comment by Tzrlk** [comment]
@vprus Indeed I've implemented that method previously. Looking through more of the issues, @angapov's requirement is answered as part of #228 with the release of [TargetGroupBinding](https://kubernetes-sigs.github.io/aws-load-balancer-controller/latest/guide/targetgroupbinding/targetgroupbinding/)s. In fact, the method I was suggesting above to hack it into place appears to have worked for one of the last commenters there.

**Comment by CPWu** [comment]
Maybe its me but it didn't work as intended for me. I created an ingress resource and it instantiated an alb, I created a secondary ingress and specified the group name and i did not see that a record was created in the hosted zone or in the rules of the alb's listener. 

`apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: hello-bye-ingress
  namespace: hello-world
  annotations:
    alb.ingress.kubernetes.io/scheme: internet-facing
    kubernetes.io/ingress.class: alb
    alb.ingress.kubernetes.io/load-balancer-name: hello-world-loadbalancer
    alb.ingress.kubernetes.io/target-type: instance
    alb.ingress.kubernetes.io/listen-ports: '[{"HTTP": 80}, {"HTTPS":443}]'
    alb.ingress.kubernetes.io/healthcheck-path: /
    alb.ingress.kubernetes.io/healthcheck-protocol: HTTP
    alb.ingress.kubernetes.io/group.name: test
spec:
  rules:
  - host: helloworld.url.test.ai
    http:
      paths: 
      - backend:
          serviceName: hello-world-service
          servicePort: 80
  - host: byeworld.url.test.ai
    http:
      paths: 
      - backend:
          serviceName: bye-world-service
          servicePort: 80`

`apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: secondary-ingress
  namespace: hello-world
  annotations:
    alb.ingress.kubernetes.io/group.name: test
spec:
  rules:
  - host: this.url.test.ai
    http:
      paths: 
      - backend:
          serviceName: this-world-service
          servicePort: 80`

**Comment by thunder-spb** [comment]
@CPWu What's in alb ingress controller logs? and what's in external-dns logs?

**Comment by marcosdiez** [comment]
To whom it may concern, I created a PR that partially solves this: https://github.com/kubernetes-sigs/aws-load-balancer-controller/pull/2655

The idea is that you still generate your ALB, the alb rules and the target group with terraform (or whatever else)
but now it's possible for the aws load balancer controller to locate the target group via it's name (which happens to be unique).

**Comment by divvy19** [comment]
I was able to resolve this using below steps :- 
Add below tags in the already existing ALB --
1)  ingress.k8s.aws/resource  --> LoadBalancer 
2)  ingress.k8s.aws/stack  --> exisitingalb
3)  elbv2.k8s.aws/cluster    --> name of your cluster

Once the above step is done add the below annotation mapping to your ingress.yaml file

alb.ingress.kubernetes.io/group.name: "exisitingalb" 

Save and apply this config using kubectl apply -f ingress.yaml.

This should use the existing load balancer as an ingress resource to provision the alb. All the magic happens via tags

**Comment by carlosrochap** [comment]
> Reference in

This didn't work for me.  If you change the ingress name the existing ALB gets deleted and a new ALB is provisioned

**Comment by captify-sjunaid** [comment]
> I was able to resolve this using below steps :- Add below tags in the already existing ALB --
> 
> 1. ingress.k8s.aws/resource  --> LoadBalancer
> 2. ingress.k8s.aws/stack  --> exisitingalb
> 3. elbv2.k8s.aws/cluster    --> name of your cluster
> 
> Once the above step is done add the below annotation mapping to your ingress.yaml file
> 
> alb.ingress.kubernetes.io/group.name: "exisitingalb"
> 
> Save and apply this config using kubectl apply -f ingress.yaml.
> 
> This should use the existing load balancer as an ingress resource to provision the alb. All the magic happens via tags

This doesn't work for me. I have tested it in ALB chart 1.4.5/2.4.4 and set the same group.name but ingress creates a new ALB instead of using the existing one created manually

**Comment by geastman3** [comment]
@divvy19 or anyone else that knows...if the ingress is deleted, does the controller try to delete the ALB? We can remove the permissions from the controller to be able to create and delete the ALB but I would also be curious if this floods the controller logs with nonstop exceptions while the controller tries to reconcile the deleted ingress object and delete the ALB?

**Comment by warnerm8** [comment]
This [blog](https://aws.amazon.com/blogs/containers/a-deeper-look-at-ingress-sharing-and-target-group-binding-in-aws-load-balancer-controller/) looks to be related:

"TargetGroupBinding is a custom resource managed by the AWS Load Balancer Controller. It allows you to expose Kubernetes applications using existing load balancers. A TargetGroupBinding resource binds a Kubernetes Service with a load balancer target group. When you create a TargetGroupBinding resource, the controller automatically configures the target group to route traffic to a Service."
